src/mame/mattel/pixtermu.cpp

```diff
@@ -0,0 +1,205 @@
+    --------
+
+    Model H4651/J4287/J4288:
+
+    - PCB Revision: PT1543A-BGA-4F2C 2005/04/30 Rev 4.2b
+    - CPU: Sharp LH79524-NOE (ARM720T)
+    - RAM: ICSI IC42S16100-7T (2MB each, 4MB total)
+    - ROM: Chip-On-Board, selected by pin CS1
+
+    JTAG
+    ----
+
+    To place the LH79524 into ARM ICE debug mode you will need to connect TEST1 via a 4.7K Ohm resistor to ground.
+
+    Pinout:
+
+    - D2: nTRST (Reset Input)
+    - P4: TMS (Mode Select Input)
+    - T3: TCK (Clock Input)
+    - T1: TDI (Serial Data Input)
+    - P3: TDO (Data Serial Output)
+    - TEST1: ARM ICE debug mode
+
+*******************************************************************************/
+
+#include "emu.h"
+
+#include "bus/generic/carts.h"
+#include "bus/generic/slot.h"
+#include "cpu/arm7/arm7.h"
+#include "cpu/arm7/arm7core.h"
+
+#include "softlist_dev.h"
+
+namespace {
+
+class pixter_multimedia_state : public driver_device
+{
+public:
+	pixter_multimedia_state(const machine_config &mconfig, device_type type, const char *tag)
+		: driver_device(mconfig, type, tag)
+		, m_cart(*this, "cartslot")
+		, m_maincpu(*this, "maincpu")
+		, m_ndcs0(*this, "ndcs0")
+		, m_internal_sram(*this, "internal_sram")
+		, m_apb(*this, "apb", 0x27000, ENDIANNESS_LITTLE)
+		, m_remap_view(*this, "remap")
+	{ }
+
+	void pixter_multimedia(machine_config &config);
+
+private:
+	// Remap Control, mapped at 0xfffe2008, offset 0x22008/4
+	static inline constexpr uint32_t APB_REMAP = 34818;
+	// Power-up Boot Configuration, mapped at 0xfffe6000, offset 0x26000/4
+	static inline constexpr uint32_t APB_PBC = 38912;
+	// nCS1 Override, mapped at 0xfffe6004, offset 0x26004/4
+	static inline constexpr uint32_t APB_CS1OV = 38913;
+	// External Peripheral Mapping, mapped at 0xfffe6008, offset 0x26008/4
+	static inline constexpr uint32_t APB_EPM = 38914;
+
+	virtual void machine_start() override;
+	virtual void machine_reset() override;
+
+	void arm7_map(address_map &map);
+	void apb_bridge_w(offs_t offset, uint32_t data, uint32_t mem_mask);
+	void apb_remap(uint32_t data);
+
+	required_device<generic_slot_device> m_cart;
+	required_device<arm7_cpu_device> m_maincpu;
+	required_shared_ptr<uint32_t> m_ndcs0;
+	required_shared_ptr<uint32_t> m_internal_sram;
+	memory_share_creator<uint32_t> m_apb;
+	memory_view m_remap_view;
+};
+
+void pixter_multimedia_state::apb_remap(uint32_t data)
+{
+	// User's Guide - 1.6 Memory Interface Architecture
+	if (data == 0) {
+		m_remap_view.select((m_apb[APB_PBC] & 0b0100) && (m_apb[APB_CS1OV] & 0b1) ? 0 : 4);
+	} else if (data < 4) {
+		m_remap_view.select(data);
+	} else {
+		logerror("Unexpected remap %d\n", data);
+	}
+}
+
+void pixter_multimedia_state::machine_start()
+{
+	if (m_cart->exists()) {
+		std::string region_tag;
+		memory_region *cart_rom = m_cart->memregion("cart:rom");
+		uint32_t const size = m_cart->common_get_size("rom");
+		m_maincpu->space(AS_PROGRAM).install_rom(0x48000000, 0x48000000 + size - 1, size, cart_rom->base());
+	}
```

Comment:
You don’t need `common_get_size` here – you can use `cart_rom->bytes()` to get the size of the region in bytes directly.

I’m not sure the mirroring specified here is going to do exactly what you want:
* For a 0x400000 byte cartridge, the mirror bits will be 0x400000, so it will be mapped at 0x48000000-0x483fffff and 0x48800000-0x487fffff.
* For a 0x200000 byte cartridge, the mirror bits will be 0x200000, so it will be mapped at 0x48000000-0x481fffff and 0x48800000-0x483fffff.

I would expect you’d want it mirrored across the same area irrespective of the cartridge size.

The `device_generic_cart_interface::install_non_power_of_two` helper is designed to help deal with this (it works for powers of two as well).  You can use it like:
```c++
		device_generic_cart_interface::install_non_power_of_two<0>(
				cart_rom->bytes(),
				0x007f'ffff,
				0,
				0x4800'0000,
				[this, cart_rom] (offs_t begin, offs_t end, offs_t mirror, offs_t src)
				{
					m_maincpu->space(AS_PROGRAM).install_rom(begin, end, mirror, cart_rom->base() + src);
				});
```

(I haven’t tested that with this driver, I’m going by memory.)

Reply:
As for what I can’t remember, it’s whether the offsets need to be shifted for the address space width or only if the address space uses shifted addresses.  If that installs at the wrong addresses, it probably needs to be:
```c++
		device_generic_cart_interface::install_non_power_of_two<2>(
				cart_rom->bytes() >> 2,
				0x007f'ffff >> 2,
				0,
				0x4800'0000,
				[this, cart_rom] (offs_t begin, offs_t end, offs_t mirror, offs_t src)
				{
					m_maincpu->space(AS_PROGRAM).install_rom(begin, end, mirror, &cart_rom->as_u32() + src);
				});
```

Reply:
First suggestion works. I rechecked on hardware and the mirroring covers the whole range up to the start of the next nCS address map, so I just needed to adjust the mask.